add blogging section using Contentlayer🚀

nkilm opened the pull request:
## Major Changes

- New blog page🎉
- All blogs are stored inside `src/content/blog` as suggested by @mxkaske 
- All routes for blogs are public - `/blog/*`

## Blog format
```md
---
title: Hello World
description: Blog using Contentlayer
author: Nikhil Mohite
authorLink: https://nikhilmohite.com
publishedAt: 2023-07-18
---

Hello World!
```

## Fixes

- #59

## Possible improvements/Additions

- The UI can be revamped to make the blog more visually appealing.
- Adding `Table of Contents` to each blog.
- New MDX components.
- Cover image for each blog - can be used as [OG](https://ogp.me/) image also.
- Add `copy` button for each code snippet.

## Known issues with [contentlayer](https://contentlayer.dev)

- https://github.com/contentlayerdev/contentlayer/issues/506

For now, the solution provided is to override the package versions in `package.json`

--- 

## Visuals

![image](https://github.com/openstatusHQ/openstatus/assets/79012023/e0970d25-0cb5-4c31-a9a8-dcbe683e303d)

![image](https://github.com/openstatusHQ/openstatus/assets/79012023/e1c68e68-e3be-4eba-9cb5-a5aba93486d3)

#### Coding snippets

Theme used - `one-dark-pro`

![image](https://github.com/openstatusHQ/openstatus/assets/79012023/2dc89bef-4b95-4fb4-b7ba-92b7582309e5)

nkilm committed:
setup contentlayer

nkilm committed:
Merge branch 'main' into add-blog

nkilm committed:
ignore output generated by contentlayer

nkilm committed:
define blog type and add plugins

nkilm committed:
add path alias for contentlayer

nkilm committed:
install mdx plugins

nkilm committed:
add sample blog

nkilm committed:
make /blog and /blog/[slug] public

nkilm committed:
reduce opacity for background

nkilm committed:
update next config(.mjs to .js) and update contentlayer

nkilm committed:
implement blogpage UI

nkilm committed:
define MDX components

nkilm committed:
define custom fuction to get only particular fields for a post

nkilm committed:
add typgraphy plugin and sample blog

nkilm committed:
improve styling for code snippets

nkilm committed:
add blog author

nkilm committed:
Merge branch 'main' into add-blog

mxkaske commented on apps/web/src/app/_components/background.tsx:
```diff
@@ -12,7 +13,7 @@ export default function Background({
   return (
     <>
-      <div className="fixed h-full w-full">
+      <div className="fixed -z-50 h-full w-full">
```
Good catch!

mxkaske commented on apps/web/src/app/blog/[slug]/page.tsx:
```diff
@@ -0,0 +1,86 @@
+  }));
+}
+
+// TODO: add OG image
```
For the OG image, you can use following link that will automatically generate the `title` and `description` inside of it: `https://openstatus.dev/api/og?title=${title}&description=${description}`.

nkilm replied:
Ah I didn't notice this API, I've added the OG image now.

mxkaske commented on apps/web/src/app/blog/[slug]/page.tsx:
```diff
@@ -0,0 +1,86 @@
+  }
+  return (
+    <>
+      <article className="relative my-5 grid grid-cols-[1fr,min(90%,100%),1fr] gap-y-8 sm:grid-cols-[1fr,min(90%,100%),1fr] sm:pt-8 md:grid md:grid-cols-[1fr,min(80%,100%),1fr] lg:grid-cols-[1fr,min(70%,100%),1fr] xl:grid-cols-[1fr,minmax(auto,240px),min(50%,100%),minmax(auto,240px),1fr] xl:gap-x-5 xl:px-0 [&>*]:col-start-2 xl:[&>*]:col-start-3">
```
You can wrap your page inside of a `Shell` component. The article doesn't have to include all the different `grid` utility classes.

mxkaske commented on apps/web/src/app/blog/[slug]/page.tsx:
```diff
@@ -0,0 +1,86 @@
+        <div>
+          <Button variant="default">
+            <Link href="/blog">⬅ Blogs</Link>
+          </Button>
```
```suggestion
          <Button variant="link" asChild>
            <Link href="/blog">⬅ Blogs</Link>
          </Button>

```

mxkaske replied:
We have the `BackButton` component that would fit here as well!

nkilm replied:
Done! I've reused `BackButton`

mxkaske commented on apps/web/src/app/page.tsx:
```diff
@@ -37,6 +37,9 @@ export default async function Page() {
                   Star on GitHub
                 </a>
               </Button>
+              <Button variant="link">
+                <Link href="/blog">Blog</Link>
+              </Button>
```
```suggestion
              <Button variant="link" asChild>
                <Link href="/blog">Blog</Link>
              </Button>
```

mxkaske commented on apps/web/src/app/blog/[slug]/page.tsx:
```diff
@@ -0,0 +1,86 @@
+          </div>
+        </section>
+
+        <main className="prose prose-p:text-sm md:prose-p:text-base prose-ul:text-sm md:prose-ul:text-base lg:prose-lg dark:prose-invert prose-code:before:content-none prose-code:after:content-none mb-6 w-full max-w-none">
```
I'd say, just use `prose dark:prose-invert` no need to adjust the font sizes for now. Keep the html tag class names though.

nkilm replied:
Okay, I've removed some of the classes. But the font size for code snippets look bit big😅

mxkaske commented on apps/web/src/app/blog/page.tsx:
```diff
@@ -0,0 +1,48 @@
+      <h1 className="text-foreground font-cal mb-4 mt-2 text-5xl">Blog</h1>
+      <Button variant="secondary">
+        <Link href="/">⬅ Back to home</Link>
+      </Button>
```
Use the `BackButton` maybe?

mxkaske commented on apps/web/src/components/Mdx.tsx:
```diff
@@ -0,0 +1,17 @@
+  return (
+    <article className="prose prose-quoteless prose-neutral dark:prose-invert max-w-none">
+      <MDXComponent components={{ ...components }} />
+    </article>
```
Here you define `prose` twice. Maybe keep that here and remove it from above.

nkilm replied:
I didn't see that, fixed it!

mxkaske commented on apps/web/src/components/MdxComponents.tsx:
```diff
@@ -0,0 +1,75 @@
+import React from "react";
```
Let's stick with the default variables for `h1`, `h2`,... and just keep the `a` tag as it is a special case.

nkilm replied:
Okay, I'll keep the default styling.

mxkaske commented on apps/web/src/components/MdxComponents.tsx:
```diff
@@ -0,0 +1,75 @@
+        {...props}
+      />
+    );
+  },
```
Great use case!

mxkaske commented on apps/web/src/lib/rehypePrettyCode.ts:
```diff
@@ -0,0 +1,21 @@
+import type { Options } from "rehype-pretty-code";
```
Can we rename the file to `rehype-pretty-code` to stick to the file name convention.

mxkaske commented on apps/web/src/lib/contentlayer.ts:
```diff
@@ -0,0 +1,38 @@
+      if (yearB !== yearA) return yearB - yearA;
+      if (monthB !== monthA) return monthB - monthA;
+      return dayB - dayA;
+    });
```
We can compare the dates via timestamps like `date1.getTime() - date2.getTime() > 0`. Removes some lines.

nkilm replied:
Great!

mxkaske commented on apps/web/src/content/types/Blog.ts:
```diff
@@ -0,0 +1,58 @@
+import { defineDocumentType } from "contentlayer/source-files";
+import readingTime from "reading-time";
+
+export const Blog = defineDocumentType(() => ({
```
Let's rename to `Post` (and the corresponding names).

nkilm replied:
Alright

mxkaske reviewed (changes requested):
Hey @nkilm! Thanks a lot for your PR!

Besides the comments, some structural things:

1. please use the "kebab-case" file name convention for all the files (e.g. `mdx-components.*`)
2. let's create a `@/lib/contentlayer` folder to put in the utility files that are used for it
3. The example post misses some attributes (`author` and `authorLink`)

What I am thinking for (2) is to:
- move `@/lib/contentlayer/rehype-pretty-code.ts` into it
- move the current `@/lib/contentlayer.ts` to `@/lib/contentlayer/blog.ts`
- move `@/content/types/Blog.ts` to `@/lib/contentlayer/documents/post.ts`

Regarding (3), I'd also suggest to have the possibility to have something like:

```
author:
  - name: Nkilm
  - profileImg: https://url.to/image
  - url: https://url.to/social/media
```

So just to group everything a bit more.

Good work! 🚀

Hit me up if you have questions. Also happy to pair on that one!

nkilm committed:
refactor blog components

Changes:
- Add OG image to each blog using /api/og
- Reuse back-button component
- Put all components related to contentlayer in separate folder
- Change file names to use kebab-case
- Use default styling for html tags used in a post

nkilm commented:
@mxkaske I've implemented the requested changes, please have a look.

mxkaske reviewed (approved):
Thanks @nkilm for the changes! I'll merge it and will small updates here and there. 

Appreciate the work! 🚀

nkilm commented:
@mxkaske Great!
